Split up zerver.lib.actions

andersk opened the pull request:
This was mostly generated by an automated splitting tool I wrote using [LibCST](https://libcst.readthedocs.io/en/latest/), with some manual `fixup!` commits (to be squashed into the automatic ones before merging). It uses the [plan](https://docs.google.com/spreadsheets/d/1ZB_hg47NL1ehC4QJ9CJPxhrs2VLbJQxpBYkLjBCx448/edit#gid=0) drafted by Tim as input.

<details>
<summary>Splitting tool</summary>

```python
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Union

import libcst as cst
import libcst.matchers as m
from libcst.codemod import CodemodContext
from libcst.codemod.visitors import RemoveImportsVisitor
from libcst.helpers import ensure_type, get_full_name_for_node_or_raise

old_file_name = "zerver/lib/actions.py"

plan = [
    line.split("\t")
    for line in (Path(__file__).parent / "plan.txt").read_text().splitlines()
]

ordered_file_names = (Path(__file__).parent / "order.txt").read_text().splitlines()


def import_from_base(mod_name: str, node: cst.ImportFrom) -> str:
    if node.relative:
        base = mod_name.rsplit(".", len(node.relative))[-len(node.relative)]
        if node.module is not None:
            base += "." + get_full_name_for_node_or_raise(node.module)
        return base
    else:
        assert node.module is not None
        return get_full_name_for_node_or_raise(node.module)


class FindImports(cst.CSTVisitor):
    def __init__(self, old_mod_name: str, this_mod_name: str) -> None:
        self.old_mod_name = old_mod_name
        self.this_mod_name = this_mod_name
        self.imports = set[str]()

    def visit_Import(self, node: cst.Import) -> None:
        for alias in node.names:
            assert alias.evaluated_name != self.old_mod_name

    def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
        base = import_from_base(self.this_mod_name, node)
        if isinstance(node.names, cst.ImportStar):
            # assert base != self.old_mod_name
            pass
        elif base == self.old_mod_name:
            self.imports.update(alias.evaluated_name for alias in node.names)
        else:
            for alias in node.names:
                assert f"{base}.{alias.evaluated_name}" != self.old_mod_name


class AdjustImports(cst.CSTTransformer):
    def __init__(
        self,
        old_mod_name: str,
        new_mod_expr: Union[cst.Attribute, cst.Name],
        this_mod_name: str,
        move_attributes: set[str],
    ) -> None:
        self.old_mod_name = old_mod_name
        self.new_mod_expr = new_mod_expr
        self.this_mod_name = this_mod_name
        self.move_attributes = move_attributes
        self.new_is_this = new_mod_expr.deep_equals(cst.parse_expression(this_mod_name))

    def leave_ImportFrom(
        self, original_node: cst.ImportFrom, updated_node: cst.ImportFrom
    ) -> Union[cst.ImportFrom, cst.FlattenSentinel[cst.ImportFrom]]:
        base = import_from_base(self.this_mod_name, updated_node)
        if base == self.old_mod_name:
            if isinstance(updated_node.names, cst.ImportStar):
                nodes = [updated_node]
                if not self.new_is_this:
                    nodes.append(updated_node.with_changes(module=self.new_mod_expr))
                return cst.FlattenSentinel(nodes)
            else:
                keep_aliases = [
                    alias.with_changes(comma=cst.Comma())
                    for alias in updated_node.names
                    if alias.name.value not in self.move_attributes
                ]
                move_aliases = [
                    alias.with_changes(comma=cst.Comma())
                    for alias in updated_node.names
                    if alias.name.value in self.move_attributes
                ]
                nodes = []
                if keep_aliases:
                    nodes.append(
                        updated_node.with_changes(
                            names=keep_aliases,
                            lpar=cst.LeftParen(),
                            rpar=cst.RightParen(),
                        )
                    )
                if move_aliases and not self.new_is_this:
                    nodes.append(
                        updated_node.with_changes(
                            module=self.new_mod_expr,
                            names=move_aliases,
                            lpar=cst.LeftParen(),
                            rpar=cst.RightParen(),
                        )
                    )
                return cst.FlattenSentinel(nodes)
        else:
            return updated_node


def strip_if_pass(mod: cst.Module) -> cst.Module:
    return mod.with_changes(
        body=[
            s
            for s in mod.body
            if not (
                isinstance(s, cst.If)
                and all(
                    isinstance(s1, cst.SimpleStatementLine)
                    and all(isinstance(s2, cst.Pass) for s2 in s1.body)
                    for s1 in s.body.body
                )
                and s.orelse is None
            )
        ]
    )


some_import = m.OneOf(m.Import(), m.ImportFrom())
empty_if = m.If(body=m.IndentedBlock(body=[]), orelse=None)


def remove_unused_imports(mod: cst.Module) -> cst.Module:
    while True:
        context = CodemodContext(wrapper=cst.MetadataWrapper(mod))
        for import_node in m.findall(mod, some_import):
            RemoveImportsVisitor.remove_unused_import_by_node(context, import_node)
        mod1 = RemoveImportsVisitor(context).transform_module(mod)
        mod = mod1.with_changes(
            body=[s for s in mod1.body if not m.matches(s, empty_if)]
        )
        if mod.deep_equals(mod1):
            break
    return mod


@dataclass
class PyFile:
    mod: cst.Module
    imports: set[str]


def main() -> None:
    old_mod_name = old_file_name.removesuffix(".py").replace("/", ".")
    old_mod_expr = cst.parse_expression(old_mod_name)
    assert isinstance(old_mod_expr, (cst.Attribute, cst.Name))

    py_file_names = set(
        (
            subprocess.check_output(["git", "ls-files", "-z", "*.py"], text=True)
            + subprocess.check_output(
                [
                    "git",
                    "--no-pager",
                    "grep",
                    "-Plz",
                    r"\A#!.*\bpython",
                    "--",
                    ".",
                    ":!*.py",
                ],
                text=True,
            )
        )
        .removesuffix("\0")
        .split("\0")
    )
    new_file_names = {new_file_name for attribute, new_file_name in plan}
    assert new_file_names <= set(ordered_file_names)

    py_files = {}
    for py_file_name in sorted(py_file_names):
        print(f"Parsing {py_file_name}")
        mod = cst.parse_module(Path(py_file_name).read_bytes())
        find_imports = FindImports(
            old_mod_name, py_file_name.removesuffix(".py").replace("/", ".")
        )
        mod.visit(find_imports)
        if (
            py_file_name == old_file_name
            or py_file_name in new_file_names
            or find_imports.imports
        ):
            py_files[py_file_name] = PyFile(mod=mod, imports=find_imports.imports)

    old_mod = py_files[old_file_name].mod

    old_attributes = set[str]()
    for s in old_mod.body:
        if isinstance(s, cst.If):
            pass
        elif isinstance(s, cst.SimpleStatementLine):
            for ss in s.body:
                if isinstance(ss, (cst.Import, cst.ImportFrom)):
                    pass
                elif isinstance(ss, cst.Assign):
                    old_attributes.update(
                        ensure_type(
                            ensure_type(target, cst.AssignTarget).target, cst.Name
                        ).value
                        for target in ss.targets
                    )
                else:
                    assert False, ss
        elif isinstance(s, (cst.FunctionDef, cst.ClassDef)):
            old_attributes.add(s.name.value)
        else:
            assert False, s

    for new_file_name in ordered_file_names:
        if new_file_name not in new_file_names:
            continue
        move_attributes = {
            attribute
            for attribute, new_file_name1 in plan
            if new_file_name1 == new_file_name
        }
        new_mod_name = new_file_name.removesuffix(".py").replace("/", ".")
        new_mod_expr = cst.parse_expression(new_mod_name)
        assert isinstance(new_mod_expr, (cst.Attribute, cst.Name))

        keep_body: list[Union[cst.SimpleStatementLine, cst.BaseCompoundStatement]] = []
        move_imports: list[
            Union[cst.SimpleStatementLine, cst.BaseCompoundStatement]
        ] = []
        move_body: list[Union[cst.SimpleStatementLine, cst.BaseCompoundStatement]] = []

        for s in old_mod.body:
            if isinstance(s, cst.If):
                keep_body.append(s)
                move_imports.append(s)
            elif isinstance(s, cst.SimpleStatementLine):
                keep_line_body: list[cst.BaseSmallStatement] = []
                move_line_imports: list[cst.BaseSmallStatement] = []
                move_line_body: list[cst.BaseSmallStatement] = []

                for ss in s.body:
                    if isinstance(ss, cst.Import):
                        assert all(
                            alias.evaluated_name != new_mod_name for alias in ss.names
                        )
                        keep_line_body.append(ss)
                        move_line_imports.append(ss)
                    elif isinstance(ss, cst.ImportFrom):
                        assert not ss.relative
                        assert ss.module is not None
                        base = get_full_name_for_node_or_raise(ss.module)
                        assert isinstance(ss.names, cst.ImportStar) or all(
                            f"{base}.{alias.evaluated_name}" != new_mod_name
                            for alias in ss.names
                        )
                        keep_line_body.append(ss)
                        if base != new_mod_name:
                            move_line_imports.append(ss)
                    elif isinstance(ss, cst.Assign):
                        (should_move,) = {
                            ensure_type(
                                ensure_type(target, cst.AssignTarget).target, cst.Name
                            ).value
                            in move_attributes
                            for target in ss.targets
                        }

                        if should_move:
                            move_line_body.append(ss)
                        else:
                            keep_line_body.append(ss)
                    else:
                        assert False, ss

                if keep_line_body:
                    keep_body.append(s.with_changes(body=keep_line_body))
                if move_line_imports:
                    move_imports.append(s.with_changes(body=move_line_imports))
                if move_line_body:
                    move_body.append(s.with_changes(body=move_line_body))
            elif isinstance(s, (cst.FunctionDef, cst.ClassDef)):
                if s.name.value in move_attributes:
                    move_body.append(s)
                else:
                    keep_body.append(s)
            else:
                assert False, s

        keep_imports = []
        if move_attributes:
            keep_imports.append(
                cst.SimpleStatementLine(
                    body=[
                        cst.ImportFrom(
                            module=new_mod_expr,
                            names=[
                                cst.ImportAlias(name=cst.Name(value=attribute))
                                for attribute in move_attributes
                            ],
                        )
                    ]
                )
            )
        if old_attributes - move_attributes:
            move_imports.append(
                cst.SimpleStatementLine(
                    body=[
                        cst.ImportFrom(
                            module=old_mod_expr,
                            names=[
                                cst.ImportAlias(name=cst.Name(value=attribute))
                                for attribute in old_attributes - move_attributes
                            ],
                        )
                    ]
                )
            )

        old_i = max(
            i
            for i, s in enumerate(keep_body)
            if isinstance(s, cst.SimpleStatementLine)
            and isinstance(s.body[0], (cst.Import, cst.ImportFrom))
        )
        old_mod = py_files[old_file_name].mod = old_mod.with_changes(
            body=[*keep_body[: old_i + 1], *keep_imports, *keep_body[old_i + 1 :]]
        )
        existing = new_file_name in py_file_names
        if existing:
            new_mod = py_files[new_file_name].mod
            new_i = max(
                i
                for i, s in enumerate(new_mod.body)
                if isinstance(s, cst.SimpleStatementLine)
                and isinstance(s.body[0], (cst.Import, cst.ImportFrom))
            )
            new_mod = new_mod.with_changes(
                body=[
                    *new_mod.body[: new_i + 1],
                    *move_imports,
                    *new_mod.body[new_i + 1 :],
                    *move_body,
                ]
            )
        else:
            assert not Path(new_file_name).exists()
            new_mod = cst.Module(body=[*move_imports, *move_body])

        old_attributes -= move_attributes

        old_mod = remove_unused_imports(old_mod)
        new_mod = remove_unused_imports(new_mod)

        Path(new_file_name).parent.mkdir(exist_ok=True)
        Path(old_file_name).write_text(old_mod.code)
        Path(new_file_name).write_text(new_mod.code)

        if existing:
            py_files[new_file_name].mod = new_mod
        else:
            find_imports = FindImports(old_mod_name, new_mod_name)
            new_mod.visit(find_imports)
            py_files[new_file_name] = PyFile(mod=new_mod, imports=find_imports.imports)

        modified = [old_file_name, new_file_name]

        for file_name, py_file in py_files.items():
            if py_file.imports & move_attributes:
                print(f"Adjusting imports in {file_name}")
                py_file.mod = py_file.mod.visit(
                    AdjustImports(
                        old_mod_name,
                        new_mod_expr,
                        file_name.removesuffix(".py").replace("/", "."),
                        move_attributes,
                    )
                )
                Path(file_name).write_text(py_file.mod.code)
                modified.append(file_name)

        print("Reformatting")
        subprocess.check_call(["black", "--", *modified])
        subprocess.check_call(["isort", "--", *modified])
        subprocess.check_call(["black", "--", *modified])
        subprocess.check_call(["git", "add", "--", *modified])
        subprocess.check_call(
            [
                "git",
                "commit",
                "-sm",
                f"actions: Move part into {new_mod_name}."
                if existing
                else f"actions: Split out {new_mod_name}.",
                "--",
                *modified,
            ]
        )


if __name__ == "__main__":
    main()
```

</details>

### To do

- [x] The plan results in three import cycles that need to be avoided:
  - [x] zerver/actions/default_streams.py ↔︎ zerver/actions/streams.py
    https://github.com/zulip/zulip/blob/d1e172dced9d7708b1a3384c54420975e5d0d799/zerver/actions/default_streams.py#L6-L10 https://github.com/zulip/zulip/blob/d1e172dced9d7708b1a3384c54420975e5d0d799/zerver/actions/streams.py#L14-L17
  - [x] zerver/actions/realm.py ↔︎ zerver/actions/realm_creation.py
    https://github.com/zulip/zulip/blob/d1e172dced9d7708b1a3384c54420975e5d0d799/zerver/actions/realm.py#L31 https://github.com/zulip/zulip/blob/d1e172dced9d7708b1a3384c54420975e5d0d799/zerver/actions/realm_creation.py#L10
  - [x] zerver/lib/streams.py ↔︎ zerver/lib/subscription_info.py
    https://github.com/zulip/zulip/blob/d1e172dced9d7708b1a3384c54420975e5d0d799/zerver/lib/streams.py#L38 https://github.com/zulip/zulip/blob/d1e172dced9d7708b1a3384c54420975e5d0d799/zerver/lib/subscription_info.py#L6
- [x] Commit ordering needs to be adjusted to avoid further intermediate import cycles as well. Currently it’s just alphabetical.
- [x] String references to `zerver.lib.actions` need to be adjusted, especially in the tests.
- [ ] When moving code into an existing module, put it in a sensible location, probably via manual fixups. Currently the script appends to the end.
- [ ] The plan has some comments that will require manual fixups. Seems reasonable to leave the rest of these for future PRs.
  - [x] `bot_owner_user_ids`: next to `active_user_ids`
  - [ ] `get_signups_stream`: But likely a bug that this exists.
  - [ ] also move zerver/lib/create_user => zerver/actions/create_user first???
    - This would create import cycles.
  - [ ] `check_change_bot_full_name`: but figure out if we can inline?
  - [ ] `get_default_subs`: but inlining would be better
  - [ ] `update_user_activity_interval`: also rename to `enqueue_user_activity_interval_update`
  - [ ] `update_user_presence`: also rename to `enqueue_user_presence_update`
  - [ ] also merged zerver/views/attachment.py => zerver/views/upload.py
  - [ ] “?” and “???”
- [x] Fix documentation about `zerver.lib.actions`.
- [x] Fix code comments about `zerver.lib.actions`.
- [x] What to do with `scripts/lib/pythonrc.py`?
  https://github.com/zulip/zulip/blob/0df4d8eb172bee9988d0a33c251368d2c2aa08ae/scripts/lib/pythonrc.py#L5
  - [x] Removed that line for now.
- [x] As a final step before merging, `git rebase --autosquash` to squash in the manual `fixup!` commits.

andersk commented:
I made these changes in the plan to avoid the circular imports.

```diff
@@ -31,7 +31,7 @@
 do_set_realm_user_default_setting	zerver/actions/realm.py
 do_deactivate_realm	zerver/actions/realm.py
 do_reactivate_realm	zerver/actions/realm.py
-do_change_realm_subdomain	zerver/actions/realm.py
+do_change_realm_subdomain	zerver/actions/realm_creation.py
 do_add_deactivated_redirect	zerver/actions/realm.py
 do_scrub_realm	zerver/actions/realm.py
 do_delete_user	zerver/actions/users.py
@@ -160,10 +160,10 @@
 do_change_default_stream_group_name	zerver/actions/default_streams.py
 do_change_default_stream_group_description	zerver/actions/default_streams.py
 do_remove_default_stream_group	zerver/actions/default_streams.py
-get_default_streams_for_realm	zerver/actions/streams.py
+get_default_streams_for_realm	zerver/actions/default_streams.py
 get_default_subs	zerver/actions/create_user.py
-streams_to_dicts_sorted	zerver/actions/streams.py
-default_stream_groups_to_dicts_sorted	zerver/actions/streams.py
+streams_to_dicts_sorted	zerver/actions/default_streams.py
+default_stream_groups_to_dicts_sorted	zerver/actions/default_streams.py
 do_update_user_activity_interval	zerver/actions/user_activity.py
 do_update_user_activity	zerver/actions/user_activity.py
 send_presence_changed	zerver/actions/presence.py
@@ -190,8 +190,8 @@
 do_update_message	zerver/actions/message_edit.py
 do_delete_messages	zerver/actions/message_edit.py
 do_delete_messages_by_sender	zerver/actions/message_edit.py
-gather_subscriptions_helper	zerver/lib/streams.py
-gather_subscriptions	zerver/lib/streams.py
+gather_subscriptions_helper	zerver/lib/subscription_info.py
+gather_subscriptions	zerver/lib/subscription_info.py
 ActivePresenceIdleUserData	zerver/actions/message_send.py
 get_active_presence_idle_user_ids	zerver/actions/message_send.py
 filter_presence_idle_user_ids	zerver/actions/message_send.py
```

timabbott commented:
I think those updates to resolve import cycles all make sense and you should feel to update them in the spreadsheet.

andersk commented:
The “also move zerver/lib/create_user => zerver/actions/create_user first???” comment on the plan would create additional import cycles, so I’m going to skip that.

https://github.com/zulip/zulip/blob/785ef8f5299e59cc11ac78c2165d0cb72d6e5ea2/zerver/actions/create_user.py#L16-L19 https://github.com/zulip/zulip/blob/785ef8f5299e59cc11ac78c2165d0cb72d6e5ea2/zerver/actions/message_send.py#L33 https://github.com/zulip/zulip/blob/785ef8f5299e59cc11ac78c2165d0cb72d6e5ea2/zerver/actions/streams.py#L19 https://github.com/zulip/zulip/blob/785ef8f5299e59cc11ac78c2165d0cb72d6e5ea2/zerver/actions/users.py#L18

timabbott commented:
I'm considering that `create_user_messages` should perhaps go to `zerver/actions/message_send.py`, leaving `zerver/lib/user_message.py` quite small. That's its only caller, it's likely to stay that way, and I think it is pretty tangled with the `SendMessageRequest` data structure (Which should probably eventually be its argument).

timabbott commented:
I think it's likely some of the stuff this moves to `streams.py` belongs in `zerver/lib/subscription_info.py`, which was created in f9ddd801221565292e298d9ea3567908a5d4f6d3.  In particular, `bulk_get_subscriber_user_ids`, `validate_user_access_to_subscribers_helper`, `validate_user_access_to_subscribers`, and `get_subscribers_query`.

timabbott commented:
We could also think about creating `zerver/lib/internal_send.py` for the internal send message batch of functions; I don't see an obvious reason they need to be in the same place. But it may be not helpful, in that most files that import them would need to import `do_send_message` or `check_send_message` anyway, so I think there's no reason to tangle this with attempting that.

timabbott commented:
Overall, this is great. https://github.com/zulip/zulip/pull/21770#issuecomment-1098590763 and https://github.com/zulip/zulip/pull/21770#issuecomment-1098593718 are the two things that I think might be worth adjusting.

andersk commented:
There’s actually no overlap between `internal_send_*` users and `do_send_message`/`check_send_message` users except for `zerver/lib/email_mirror.py`. I’ll see what happens if I split those.

andersk commented:
Oh, this call cycle means that’s not happening.

`internal_send_private_message` →
`internal_prep_private_message` → 
`_internal_prep_message` →
`check_message` →
`validate_stream_{id,name}_with_pm_notification` →
`send_pm_if_empty_stream` →
`send_rate_limited_pm_notification_to_bot_owner` →
`internal_send_private_message`.

andersk commented:
Moved `create_user_messages`, `validate_user_access_to_subscribers`, `validate_user_access_to_subscribers_helper`, `bulk_get_subscriber_user_ids`, `get_subscribers_query`.

timabbott commented:
I think all of those moves look to be an improvement! I did another pass and have the following thoughts:

`message_send.py` could use some reordering:
* `create_user_messages` seems like it should be above `do_send_messages`, rather than just after, given that it's a helper used at the top of `do_send_messages`?
* `filter_presence_idle_user_ids` should be before `get_active_presence_idle_user_ids`, and those two should also not be right at the end but instead above where they're called with the similar helpers. That would leave the `internal_send` stuff as a block at the end, which feels nice.

We should look at the order for the other two really big files since this is a free function reordering opportunity:
* I think `message_edit.py` is mostly fine, but perhaps we should have `check_update_message` be below `do_update_message`, which it calls? I feel the current order is more defensible in that it's potentially nice to have the the error-checking logic at the top, so one sees it first if reading the file, so I'm fine with leaving this as-is if you like the order.
* In `zerver/actions/streams.py`, I think `bulk_add_subscriptions` should be moved after `send_peer_subscriber_events`, since the 3 functions we'd be moving past are all helpers that we extracted from `bulk_add_subscriptions` itself in the past.  I might all move `do_deactivate_stream` to the end, just before `do_get_streams`.
* I'm not sure `do_get_streams` really belongs in `actions` code at all; it's not a write operation. If we can get it into `zerver/lib/streams.py` without creating import cycles, that's probably the most natural home for it (in that similar "Bulk fetch object X from the database with these query parameters" live in that class of file, e.g. in `zerver/lib/users.py`).

Finally a couple naming questions:
* Should `zerver/actions/realm.py` be called `zerver/actions/realm_settings.py` instead? I think everything in there is a realm setting that modifies the `Realm` object, so both are potentially defensible. 
* Should `zerver/actions/realm_creation.py` be `zerver/actions/create_realm.py` instead, to make it sort nicely next to the closely related `zerver/actions/create_user.py`? I mostly feel like `realm_creation` is not a type of name we use; `realm_create` would also be fine in my view.

andersk commented:
Made all of those changes.

* Reordered `zerver.actions.message_send` in b82197c06863172859d8c87aeaeb8ed20ecea7a3 (I have also moved `class ActivePresenceIdleUserData`).
* Reordered `zerver.actions.message_edit` in 91fc3b596c34f5e0ec91306dadc670a98d0e4ce3.
* Reordered `zerver.actions.streams` in 4eca376cd4a7f0aded142d5c2d228e8c6c3d2c0c (I have also moved `SubT`.)
* Moved `do_get_streams` to `zerver/lib/streams.py`.
* Renamed `zerver/actions/realm.py` to `zerver/actions/realm_settings.py`.
* Renamed `zerver/actions/realm_creation.py` to `zerver/actions/create_realm.py`. (Note, the previous name came from you on the plan spreadsheet; I think I’ve done no editorializing myself in this PR.)

timabbott commented:
Those changes all look good!  (And yeah, I realize that the above things were likely my original plan -- just details I clearly ).

You should feel free to do the squash and merge step at this point, and then to make a backported version for the 5.x branch so that we can still backport changes that involve `actions.py`.
